nixos/modules/security/permissions-wrappers/permissions-wrapper.c

```diff
@@ -0,0 +1,224 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <assert.h>
+#include <errno.h>
+
+// Make sure assertions are not compiled out, we use them to codify
+// invariants about this program and we want it to fail fast and
+// loudly if they are violated.
+#undef NDEBUG
+
+extern char **environ;
+
+// The SOURCE_PROG and WRAPPER_DIR macros are supplied at compile time
+// for a security reason: So they cannot be changed at runtime.
+static char * sourceProg = SOURCE_PROG;
+static char * wrapperDir = WRAPPER_DIR;
+
+// Make sure we have the WRAPPER_TYPE macro specified at compile
+// time...
```

Comment:
@7c6f434c I think I'm confused by your question:

> Why not have a single binary to copy as needed?

Are you asking why are we producing a binary-per wrapper instead of having a single? It felt safer to me to statically specify the source program path instead of reading it from a companion `.real` file. It also feels more idiomatic to Nix IMHO.

I think I had more reasons when I wrote it too but can't remember them all now. The above justification is the one I remember the clearest though.

> So, what will setcap wrapper do wrong if it is used as setuid wrapper?

I'm confused, could you rephrase your question?